### PR TITLE
feat: add maximum concurrent threads by using Parallel gem

### PR DESCRIPTION
**Pull request by RickCSong**
Before, asset_sync would create a separate thread for every asset that
was uplaoded. When there are a large number of assets being uploaded,
this could lead to processes crashing due to too many threads being
created.

By limiting the number of threads, this speeds up
performance while preventing crashes from resource starvation. 

See: https://github.com/AssetSync/asset_sync/issues/395

### CODE REVIEW

**Comment by RickCSong** [comment]
This is really great feedback @PikachuEXE ! I've updated the PR accordingly -- great callout on using `.close`.

File: lib/asset_sync/storage.rb [hunk 1]
```diff
@@ -248,19 +248,29 @@ def upload_files
...
+        jobs.close
+
+        num_threads = [self.config.concurrent_uploads_max_threads, local_files_to_upload.length].min
         # Upload new files
-        local_files_to_upload.each do |f|
-          next unless File.file? "#{path}/#{f}" # Only files.
-          threads.add(Thread.new { upload_file f })
+        workers = Array.new(num_threads) do
+          Thread.new do
+            begin
```
Comment (PikachuEXE) [inline lib/asset_sync/storage.rb]: Why do you still need `begin..end` in `Thread` block?

Comment (RickCSong) [reply lib/asset_sync/storage.rb]: Good catch!!! Fixed.

**Comment by PikachuEXE** [comment]
@RickCSong 
You forgot to solve the conflict again lol

**Comment by RickCSong** [comment]
Sorry @PikachuEXE ! Done.

**Comment by PikachuEXE** [comment]
Released in `2.10.0`

**Pull request merged**